### PR TITLE
fix record batch writing to dataset

### LINKED ISSUE #1019: Alice Bob example fails
The example that shows how to create data from record batch fails with the following error

```
import lance
import pyarrow as pa

def producer():
    yield pa.RecordBatch.from_pylist([{"name": "Alice", "age": 20}])
    yield pa.RecordBatch.from_pylist([{"name": "Blob", "age": 30}])


lance.write_dataset(producer, "./alice_and_bob.lance")
```

```
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 lance.write_dataset(producer, "[./alice_and_bob.lance](https://file+.vscode-resource.vscode-cdn.net/Users/jrabarisoa/Codes/metrics/alice_and_bob.lance)")

File [/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/site-packages/lance/dataset.py:919](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/site-packages/lance/dataset.py:919), in write_dataset(data_obj, uri, schema, mode, max_rows_per_file, max_rows_per_group)
    888 def write_dataset(
    889     data_obj: ReaderLike,
    890     uri: Union[str, Path],
   (...)
    894     max_rows_per_group: int = 1024,
    895 ) -> LanceDataset:
    896     """Write a given data_obj to the given uri
    897 
    898     Parameters
   (...)
    917 
    918     """
--> 919     reader = _coerce_reader(data_obj)
    920     # TODO add support for passing in LanceDataset and LanceScanner here
    922     params = {
    923         "mode": mode,
    924         "max_rows_per_file": max_rows_per_file,
    925         "max_rows_per_group": max_rows_per_group,
    926     }

File [/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/site-packages/lance/dataset.py:947](https://file+.vscode-resource.vscode-cdn.net/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/site-packages/lance/dataset.py:947), in _coerce_reader(data_obj, schema)
    945     return data_obj
    946 else:
--> 947     raise TypeError(f"Unknown data_obj type {type(data_obj)}")

TypeError: Unknown data_obj type
```

### PR DESCRIPTION
**Pull request by TevinWang**
Implement writing dataset with `Iterator[RecordBatch]`, like mentioned in the docs. This requires an argument of schema, so we throw a ValueError if it doesn't exist.

Minor ugliness: I check for `Iterator` type, not `Iterator[RecordBatch]`. Let me know if this causes some concerns.

Closes #1019

### CODE REVIEW

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -942,6 +945,13 @@ def _coerce_reader(
         return pa.dataset.Scanner.from_dataset(data_obj).to_reader()
     elif isinstance(data_obj, pa.dataset.Scanner):
         return data_obj.to_reader()
+    elif isinstance(data_obj, Iterator):
+        if schema is not None:
+            return pa.RecordBatchReader.from_batches(schema, data_obj)
+        else:
```
Comment (changhiskhan) [inline python/python/lance/dataset.py:951]: Can we use the schema from the first batch?

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -942,6 +945,13 @@ def _coerce_reader(
         return pa.dataset.Scanner.from_dataset(data_obj).to_reader()
     elif isinstance(data_obj, pa.dataset.Scanner):
         return data_obj.to_reader()
+    elif isinstance(data_obj, Iterator):
```
Comment (changhiskhan) [inline python/python/lance/dataset.py:948]: add unit test please

**Review by changhiskhan — COMMENTED** [review]
Also what about an Iterable/Sequence (e.g., list of RecordBatch)

Also worth improving the error message here: https://github.com/lancedb/lance/pull/1030/files#diff-3c0e591b6a64801b9e3f92380ab99c1546e03faaa769b6de4b37af58d95c588cR958 

It just says "Uknown data_obj type" which as you can see from #1019 it's pretty cryptic

File: python/python/lance/dataset.py (hunk 1 above)
Comment (TevinWang) [reply python/python/lance/dataset.py:951]: From what I recall, we can't access that batch anymore if we expose it in the iterator right? Let me know if I'm misunderstanding.

Comment (wjones127) [reply python/python/lance/dataset.py:951]: Yeah we'd have to implement a peekable iterator. I don't think we need to do that for now. I think beginners will simply write tables, and advanced users won't have much issue providing the schema themselves.

**Review by changhiskhan — APPROVED** [review]
Thanks @TevinWang

**Comment by TevinWang** [comment]
@changhiskhan rerequesting review here, I changed some of the logic to include iterables, but needed to move the conditionals since a RecordBatchReader is an iterable (also added unit tests). Let me know if there are any more changes.

**Review by changhiskhan — APPROVED** [review]
the new unit tests are almost identical. please collapse this into a single method

**Pull request merged**